lib/queue.js

```diff
@@ -26,6 +26,45 @@ function Queue( subQueues ) {
 
 util.inherits( Queue, events.EventEmitter );
 
+/**
+ * Create a new sub-queue.
+ * @param {String}   name  The sub-queue to create
+ * @param {String}   [before]  Add the new sub-queue before the this sub-queue.
+ *                             Otherwise the new sub-queue will be added last. 
+ */
+
+Queue.prototype.addSubQueue = function( name, before ) {
+  if ( this.__queues__[name] ) {
+    // Sub-queue already exists
+    return;
+  }
+
+  if ( !before ) {
+    // Add at last place.
+    this.__queues__[name] = new SubQueue();
+    this.queueNames.push( name );
+    return
+  }
+
+  if ( !this.__queues__[before] || _.indexOf(this.queueNames, before ) === -1 ) {
+    throw new Error('sub-queue ' + before + ' not found');
+  }
+
+  var current = this.__queues__;
+  var currentNames = Object.keys( current );
+  // Recreate the queue with new order.
+  this.__queues__ = {};
+  currentNames.forEach(function( currentName ) {
+    if ( currentName === before ) {
+      this.__queues__[name] = new SubQueue();
+    }
+    this.__queues__[currentName] = current[currentName];
+  }.bind(this));
+
+  // Recreate queueNames
+  this.queueNames = Object.keys( this.__queues__ );
```

Comment:
I think this logic can be simplified a lot using `splice` and injecting the new queue at the right position.

Theoretically, objects keys in JS aren't ordered. So it'd be better not to rely on this informal behavior. But also I think the implementation would be much simpler using splice.

Reply:
Yes, don’t like this too.
I was keeping the same behavior.
https://github.com/SBoudrias/grouped-queue/blob/128594ae1500f66b22695fdc96c0012685411d3d/lib/queue.js#L77

But I agree it needs to be changed.